library/src/scala/reflect/TypeTest.scala

```diff
@@ -0,0 +1,33 @@
+package scala.reflect
+
+/** A `TypeTest[S, T] contains the logic needed to know at runtime if a value of
+ *  type `S` can be downcased to `T`.
+ *
+ *  If a pattern match is performed on a term of type `s: S` that is uncheckable with `s.isInstanceOf[T]` and
+ *  the pattern are of the form:
+ *    - `t: T`
+ *    - `t @ X()` where the `X.unapply` has takes an argument of type `T`
+ *  then a given instance of `TypeTest[S, T]` is summoned and used to performed the test.
+ *
+ *  Note: This is replacemet for `ClassTag.unapply` that can be sound for path dependent types
+ */
+@scala.annotation.implicitNotFound(msg = "No TypeTest available for [${S}, ${T}]")
+trait TypeTest[-S, T] extends Serializable {
+
+  /** A TypeTest[S, T] can serve as an extractor that matches only S of type T.
+   *
+   * The compiler tries to turn unchecked type tests in pattern matches into checked ones
+   * by wrapping a `(_: T)` type pattern as `tt(_: T)`, where `ct` is the `TypeTest[S, T]` instance.
+   * Type tests necessary before calling other extractors are treated similarly.
+   * `SomeExtractor(...)` is turned into `tt(SomeExtractor(...))` if `T` in `SomeExtractor.unapply(x: T)`
```

Comment:
That's not accurate, is it? Shouldn't the `tt` be used first?

Reply:
`tt(SomeExtractor(...)) it is accurate. But as it is a pattern, the outer most pattern is applied first at runtime.

```scala
 x match { case tt(SomeExtractor(...)) =>  ... }
```

becomes something like
```scala
val x$1 = tt.unapply(x)
if (x$1.isDefined) {
 val x$2 = x$1.get
 val x$3 = SomeExtractor.unapply(x$2)
 if (x$3.isDefined) ...
} else ...
```